tensorrt_llm/_torch/attention_backend/trtllm.py

```diff
@@ -1059,14 +1077,76 @@ def prepare_context_mla_with_cached_kv(self,
         self.ctx_kv_indptr[:self.num_contexts + 1].copy_(
             self.host_ctx_kv_indptr[:self.num_contexts + 1], non_blocking=True)
 
+    def compute_max_num_custom_mask_tiles_kv_upper_bound(
+            self, max_seq_len_kv, min_first_sparse_mask_offset_kv,
+            tile_size_kv_per_cta) -> int:
+        """
+        Compute the conservative upper bound of numCustomMaskTilesKv.
+
+        Args:
+            max_seq_len_kv (int): The maximum seqLenKv in the batch
+            min_first_sparse_mask_offset_kv (int): The minimum firstSparseMaskOffsetKv in the batch
+            tile_size_kv_per_cta (int): tileSizeKvPerCta value
+        """
+        first_sparse_tile_offset = min_first_sparse_mask_offset_kv // tile_size_kv_per_cta
+        num_tiles_kv_total = math.ceil(max_seq_len_kv / tile_size_kv_per_cta)
+        max_num_custom_mask_tiles_kv = num_tiles_kv_total - first_sparse_tile_offset
+        return max_num_custom_mask_tiles_kv
+
+    def spec_decoding_param_prepare_for_blackwell(self) -> None:
+        """
+        Prepare the blackwell parameters for the speculative decoding (Medusa and Eagle) generation-phase attention kernels.
+        """
+        self.spec_decoding_bl_tree_mask_offset = torch.zeros(
+            [self.max_num_requests],
+            dtype=torch.int64,
+            device='cuda',
+        )
+        max_kv_len = self.kv_lens[:self.num_seqs].max()
+        assert self.kv_lens_cuda[:self.
+                                 num_seqs] >= self._seq_lens_cuda[:self.
+                                                                  num_seqs], "kv_lens should be greater than seq_lens,please run prepare() first"
+
+        # Only support seq_lens are equal in one batch
+        seq_lens_slice = self.seq_lens[:self.num_seqs]
+        assert seq_lens_slice.min() == seq_lens_slice.max(), \
+            f"All elements in seq_lens must be equal in one batch, but got min={seq_lens_slice.min()}, max={seq_lens_slice.max()}"
+
+        self.spec_bl_tree_first_sparse_mask_offset_kv = (
+            self.kv_lens_cuda[:self.num_seqs] -
+            self._seq_lens_cuda[:self.num_seqs]).to(torch.int32)
+        min_first_sparse_mask_offset_kv = self.spec_bl_tree_first_sparse_mask_offset_kv.min(
+        )
+        # tile_size_kv * tile_size_q * num_instances_q * num_instances_kv is the largest value that is used in the trtllm-gen kernels
```

Comment:
probably we can expose this a torch op from trtllm-gen fmha runner, so we have a consistent interface, but let us do in later MRs. No need for this MR. Thanks.